src/main/java/uk/gov/hmcts/reform/sscs/service/CcdNotificationsPdfService.java

```diff
@@ -79,6 +80,49 @@ public SscsCaseData mergeCorrespondenceIntoCcd(SscsCaseData sscsCaseData, Corres
         return caseDetails.getData();
     }
 
+    /**
+     * This method generates PDF using HTML, stores in doc store and updates SSCS case data retrieved from DB with correspondence
+     *
+     * @param caseId - CCD case id
+     * @param correspondence - Correspondence which needs to be added to the case data
+     */
+    public void mergeCorrespondenceIntoCcdV2(Long caseId, Correspondence correspondence) {
+        Map<String, Object> placeholders = new HashMap<>();
+        placeholders.put("body", correspondence.getValue().getBody());
+        placeholders.put("subject", correspondence.getValue().getSubject());
+        placeholders.put("sentOn", correspondence.getValue().getSentOn());
+        placeholders.put("from", correspondence.getValue().getFrom());
+        placeholders.put("to", correspondence.getValue().getTo());
+
+        byte[] template;
+        try {
+            template = getSentEmailTemplate();
+        } catch (IOException e) {
+            throw new PdfGenerationException("Error getting template", e);
+        }
+
+        byte[] pdf = pdfServiceClient.generateFromHtml(template, placeholders);
+        String filename = String.format("%s %s.pdf", correspondence.getValue().getEventType(), correspondence.getValue().getSentOn());
+        List<SscsDocument> pdfDocuments = pdfStoreService.store(pdf, filename, correspondence.getValue().getCorrespondenceType().name());
+        final List<Correspondence> correspondences = pdfDocuments.stream().map(doc ->
+                correspondence.toBuilder().value(correspondence.getValue().toBuilder()
+                        .documentLink(doc.getValue().getDocumentLink())
+                        .build()).build()
+        ).toList();
+
+        IdamTokens idamTokens = idamService.getIdamTokens();
+        final SscsCaseDetails sscsCaseDetails = ccdService.getByCaseId(caseId, idamTokens);
```

Comment:
This is not concurrency safe since this data could be stale - you should move all this inside the lambda or (a separate method) and use only the SscsCaseData passed to the lambda (which comes from the CCD start event so won't be stale).

This shows how easy it is to introduce subtle concurrency bugs and why this is such a widespread problem.

Reply:
have updated the logic now @banderous